// @haelp/teto/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CH () {
        return _channel.CH;
    },
    get ChannelAPI () {
        return _channel.ChannelAPI;
    },
    get Classes () {
        return _classes;
    },
    get Client () {
        return _classes.Client;
    },
    get Engine () {
        return _engine;
    },
    get Types () {
        return _types;
    },
    get Utils () {
        return _utils;
    },
    get ch () {
        return _channel.ch;
    },
    get version () {
        return _utils.version;
    }
});
const _utils = /*#__PURE__*/ _interop_require_wildcard(require("./utils"));
const _chalk = /*#__PURE__*/ _interop_require_default(require("chalk"));
const _channel = require("./channel");
const _types = /*#__PURE__*/ _interop_require_wildcard(require("./types"));
const _classes = /*#__PURE__*/ _interop_require_wildcard(require("./classes"));
const _engine = /*#__PURE__*/ _interop_require_wildcard(require("./engine"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const suppressKey = "TRIANGLE_VERSION_SUPPRESS";
const MIGRATION_GUIDE_URL = "https://triangle.haelp.dev/documents/Migration.html";
const parseSemver = (v)=>{
    const [major = 0, minor = 0, patch = 0] = v.split(".").map(Number);
    return {
        major,
        minor,
        patch
    };
};
const isNewer = (a, b)=>a.major > b.major || a.major === b.major && a.minor > b.minor || a.major === b.major && a.minor === b.minor && a.patch > b.patch;
if (typeof process !== "undefined" && !(suppressKey in process.env)) {
    fetch("https://registry.npmjs.org/@haelp/teto").then((r)=>r.json()).then((d)=>{
        const latest = d["dist-tags"].latest;
        const currentV = parseSemver(_utils.version);
        const latestV = parseSemver(latest);
        if (!isNewer(latestV, currentV)) return;
        if (latestV.major > currentV.major) {
            console.log(`${_chalk.default.redBright("[Triangle.js]")}: Your triangle.js version (${_utils.version}) is ${_chalk.default.bold("one or more major versions behind")} the latest release (${latest}).\n` + `This update includes breaking changes.\n` + `Migration guide: ${_chalk.default.cyan(MIGRATION_GUIDE_URL)}\n\n` + `To suppress this warning, set the ${suppressKey} environment variable.`);
        } else {
            console.log(`${_chalk.default.redBright("[Triangle.js]")}: Your triangle.js is out of date (${_utils.version} vs ${latest}). ` + `We recommend updating with 'npm install @haelp/teto@latest'.\n` + `To suppress this warning, set the ${suppressKey} environment variable.`);
        }
    }).catch(()=>{
    /* fail silently */ });
}
