// @haelp/teto/dist/engine/queue/index.mjs
import { rngMap } from "./rng/index.mjs";
export class Queue extends Array {
    seed;
    type;
    bag;
    _minLength;
    repopulateListener = null;
    static get [Symbol.species]() {
        return Array;
    }
    constructor(options){
        super();
        this.seed = options.seed;
        this.type = options.type;
        this.reset();
        this.minLength = options.minLength;
    }
    reset() {
        this.bag = new rngMap[this.type](this.seed);
        this.#repopulate();
    }
    onRepopulate(listener) {
        this.repopulateListener = listener;
    }
    get minLength() {
        return this._minLength;
    }
    set minLength(val) {
        this._minLength = val;
        this.#repopulate();
    }
    get next() {
        return this[0];
    }
    shift() {
        const val = super.shift();
        this.#repopulate();
        return val;
    }
    #repopulate() {
        const added = [];
        while(this.length < this.minLength){
            const newValues = this.bag.next();
            this.push(...newValues);
            added.push(...newValues);
        }
        if (this.repopulateListener && added.length) {
            this.repopulateListener(added);
        }
    }
    snapshot() {
        return {
            value: Array.from(this),
            bag: this.bag.snapshot()
        };
    }
    fromSnapshot(snapshot) {
        this.bag.fromSnapshot(snapshot.bag);
        this.splice(0, this.length, ...snapshot.value);
    }
    raw() {
        return Array.from(this);
    }
}
export * from "./rng/index.mjs";
export * from "./types.mjs";
