// @haelp/teto/dist/utils/events.mjs
export class EventEmitter {
    #listeners;
    #maxListeners = {
        default: 10,
        overrides: new Map()
    };
    /** Enables more debugging logs for memory leaks */ verbose = false;
    constructor(){
        this.#listeners = [];
    }
    on(event, cb) {
        this.#listeners.push([
            event,
            cb,
            false
        ]);
        const listeners = this.#listeners.filter(([e])=>e === event);
        if (listeners.length > (this.#maxListeners.overrides.get(event) ?? this.#maxListeners.default)) {
            console.warn(`Max listeners exceeded for event "${String(event)}". Current: ${this.#listeners.filter(([e])=>e === event).length}, Max: ${this.#maxListeners.overrides.get(event) ?? this.#maxListeners.default}`);
            if (this.verbose) console.warn(`Trace: ${new Error().stack}\n\nListeners:\n`, listeners.map(([_, fn])=>fn.toString()).join("\n\n"));
        }
        return this;
    }
    off(event, cb) {
        this.#listeners = this.#listeners.filter(([e, c])=>e !== event || c !== cb);
        return this;
    }
    emit(event, data) {
        const toRemove = new Set();
        const listeners = [
            ...this.#listeners
        ];
        listeners.forEach(([e, cb, once], idx)=>{
            if (e !== event) return;
            cb(data);
            if (once) toRemove.add(idx);
        });
        this.#listeners = this.#listeners.filter((_, idx)=>!toRemove.has(idx));
        return this;
    }
    once(event, cb) {
        this.#listeners.push([
            event,
            cb,
            true
        ]);
        return this;
    }
    removeAllListeners(event) {
        if (event) {
            this.#listeners = this.#listeners.filter(([e])=>e !== event);
        } else {
            this.#listeners = [];
        }
    }
    setMaxListeners(eventOrN, n) {
        const count = n ?? eventOrN;
        if (!Number.isInteger(count) || count <= 0) {
            throw new RangeError("Max listeners must be a positive integer");
        }
        if (typeof eventOrN === "number") {
            this.#maxListeners.default = eventOrN;
        } else if (Array.isArray(eventOrN)) {
            eventOrN.forEach((event)=>{
                this.#maxListeners.overrides.set(event, count);
            });
        } else {
            this.#maxListeners.overrides.set(eventOrN, count);
        }
    }
    get maxListeners() {
        return this.#maxListeners;
    }
    /**
   * @internal
   */ set _maxListeners(data) {
        this.#maxListeners = data;
    }
    export() {
        return {
            listeners: this.#listeners.map(([event, cb, once])=>({
                    event,
                    cb,
                    once
                })),
            maxListeners: this.#maxListeners,
            verbose: this.verbose
        };
    }
    import(data) {
        data.listeners.forEach(({ event, cb, once })=>{
            if (once) {
                this.once(event, cb);
            } else {
                this.on(event, cb);
            }
        });
        this.#maxListeners = data.maxListeners;
        this.verbose = data.verbose;
        return this;
    }
}
