// @haelp/teto/dist/utils/theorypack.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "pack", {
    enumerable: true,
    get: function() {
        return pack;
    }
});
const _msgpackr = /*#__PURE__*/ _interop_require_default(require("./msgpackr"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const msgpackr = (0, _msgpackr.default)();
// require theorypack extensions
msgpackr.addExtension({
    Class: undefined,
    type: 1,
    read: (e)=>null === e ? {
            success: true
        } : {
            success: true,
            ...e
        }
});
msgpackr.addExtension({
    Class: undefined,
    type: 2,
    read: (e)=>null === e ? {
            success: false
        } : {
            success: false,
            error: e
        }
});
const unpacker = new msgpackr.Unpackr({
    int64AsType: "number",
    bundleStrings: true,
    sequential: false
});
const packer = new msgpackr.Packr({
    int64AsType: "number",
    bundleStrings: true,
    sequential: false
});
(function(pack) {
    pack.unpack = unpacker.unpack.bind(unpacker);
    pack.unpackMultiple = unpacker.unpackMultiple.bind(unpacker);
    pack.decode = packer.decode.bind(unpacker);
    pack.pack = packer.pack.bind(packer);
    pack.encode = packer.encode.bind(packer);
})(pack || (pack = {}));
var pack;
