// @haelp/teto/dist/classes/social/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Social", {
    enumerable: true,
    get: function() {
        return Social;
    }
});
const _engine = require("../../engine");
const _relationship = _export_star(require("./relationship"), exports);
const _chalk = /*#__PURE__*/ _interop_require_default(require("chalk"));
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
_export_star(require("./types"), exports);
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const processRelationship = (r, selfID)=>({
        ...r,
        user: {
            id: r.from._id === selfID ? r.to._id : r.from._id,
            username: r.from._id === selfID ? r.to.username : r.from.username,
            avatar: r.from._id === selfID ? r.to.avatar_revision : r.from.avatar_revision
        }
    });
class Social {
    #client;
    /** The current number of people online */ online;
    /** List of the client's friends */ friends;
    /** List of "pending" relationships (shows in `other` tab on TETR.IO) */ other;
    /** people you block */ blocked;
    /** Notifications */ notifications;
    /** Social config */ config;
    static defaultConfig = {
        suppressDMErrors: false,
        autoLoadDMs: true,
        autoProcessNotifications: true
    };
    /** @hideconstructor */ constructor(client, config, init){
        this.#client = client;
        this.config = _lodash.default.merge(Social.defaultConfig, config);
        this.#init();
        this.online = init.online;
        if ("config" in init) {
            this.friends = init.friends.map((snapshot)=>_relationship.Relationship.fromSnapshot(snapshot, this, client));
            this.other = init.others.map((snapshot)=>_relationship.Relationship.fromSnapshot(snapshot, this, client));
        } else {
            this.friends = init.friends.map((r)=>new _relationship.Relationship({
                    id: r.user.id,
                    relationshipID: r._id,
                    username: r.user.username,
                    avatar: r.user.avatar ?? undefined
                }, this, this.#client));
            this.other = init.other.map((r)=>new _relationship.Relationship({
                    id: r.user.id,
                    relationshipID: r._id,
                    username: r.user.username,
                    avatar: r.user.avatar ?? undefined
                }, this, this.#client));
        }
        this.blocked = init.blocked;
        this.notifications = init.notifications;
    }
    static async create(client, config, initData) {
        const rel = initData.relationships.map((r)=>processRelationship(r, client.user.id));
        const data = {
            online: initData.total_online,
            notifications: initData.notifications,
            friends: rel.filter((r)=>r.type === "friend"),
            other: rel.filter((r)=>r.type === "pending"),
            blocked: rel.filter((r)=>r.type === "block").map((r)=>({
                    id: r.user.id,
                    username: r.user.username,
                    avatar: r.user.avatar ?? undefined
                }))
        };
        return new Social(client, config, data);
    }
    get #api() {
        return this.#client.api;
    }
    #init() {
        if (this.config.autoProcessNotifications) {
            setTimeout(()=>{
                this.notifications.forEach((n)=>{
                    if (!n.seen) {
                        if (n.type === "friend") {
                            const rel = processRelationship(n.data.relationship, this.#client.user.id);
                            this.#client.emit("client.friended", {
                                id: rel.user.id,
                                name: rel.user.username,
                                avatar: rel.user.avatar
                            });
                        }
                    }
                });
                this.markNotificationsAsRead();
            }, 0);
        }
        this.#client.on("social.online", (count)=>{
            this.online = count;
        });
        this.#client.on("social.notification", async (n)=>{
            this.notifications.splice(0, 0, n);
            if (this.config.autoProcessNotifications) {
                if (n.type === "friend") {
                    const rel = processRelationship(n.data.relationship, this.#client.user.id);
                    this.#client.emit("client.friended", {
                        id: rel.user.id,
                        name: rel.user.username,
                        avatar: rel.user.avatar
                    });
                    const user = this.get({
                        id: rel.user.id
                    });
                    if (!user) {
                        this.other.push(new _relationship.Relationship({
                            id: rel.user.id,
                            username: rel.user.username,
                            avatar: rel.user.avatar ?? undefined,
                            relationshipID: ""
                        }, this, this.#client));
                    }
                }
                this.markNotificationsAsRead();
            }
        });
        this.#client.on("social.dm", async (raw)=>{
            const dm = {
                ...raw,
                ts: new Date(raw.ts)
            };
            let target = dm.data.user;
            if (dm.data.user === this.#client.user.id) {
                const userID = dm.stream.split(":").filter((id)=>id !== this.#client.user.id)[0];
                // failsafe
                if (!userID) return;
                target = userID;
            }
            let user = this.get({
                id: target
            });
            if (user) {
                if (!user.dmsLoaded && this.config.autoLoadDMs) await user.loadDms();
                else user.dms.push({
                    ...dm,
                    ts: new Date(dm.ts)
                });
            } else {
                try {
                    const u = await this.who(target);
                    this.other.push(new _relationship.Relationship({
                        id: u._id,
                        username: u.username,
                        avatar: u.avatar_revision,
                        relationshipID: ""
                    }, this, this.#client));
                    user = this.get({
                        id: target
                    });
                    if (this.config.autoLoadDMs) await user.loadDms();
                } catch  {
                /* empty */ }
            }
            if (!user) {
                console.warn(`${_chalk.default.yellowBright("[Triangle.js]")}: Failed to load user data for ${target}. 'client.dm' event discareded.`);
                user = this.get({
                    id: dm.data.user
                });
                if (this.config.autoLoadDMs) await user.loadDms();
            }
            // don't trigger client.dm for messages sent by the client itself
            if (dm.data.user === this.#client.user.id) return;
            this.#client.emit("client.dm", {
                relationship: user,
                raw: dm,
                content: dm.data.content,
                reply: (content)=>this.dm(dm.data.user, content)
            });
        });
    }
    /**
   * Marks all notifications as read
   * @example
   * client.social.markNotificationsAsRead();
   */ markNotificationsAsRead() {
        this.#client.emit("social.notification.ack");
        this.notifications.forEach((n)=>n.seen = true);
    }
    get(target) {
        if (typeof target === "string") return this.friends.find((r)=>r.id === target || r.username === target) || this.other.find((r)=>r.id === target || r.username === target) || null;
        else if ("id" in target) return this.friends.find((r)=>r.id === target.id) || this.other.find((r)=>r.id === target.id) || null;
        else return this.friends.find((r)=>r.username === target.username) || this.other.find((r)=>r.username === target.username) || null;
    }
    /**
   * Get the user id given a username
   */ async resolve(username) {
        return await this.#api.users.resolve(username);
    }
    async who(id) {
        return this.#api.users.get({
            id
        });
    }
    /**
   * Send a message to a specified user (based on id).
   * Returns `Promise<"you.fail" | "they.fail" | "you.ban" | "they.ban">` if `suppressDMErrors` is `true` and an error (`social.dm.fail`) is thrown.
   * @example
   * await client.social.dm(await client.social.resolve('halp'), 'what\'s up?');
   */ async dm(userID, message) {
        try {
            const res = await this.#client.wrap("social.dm", {
                recipient: userID,
                msg: message
            }, "social.dm", [
                "social.dm.fail",
                "client.error"
            ]);
            return {
                ...res,
                ts: new Date(res.ts)
            };
        } catch (e) {
            if (this.config.suppressDMErrors) return e;
            throw e;
        }
    }
    /**
   * Send a user a friend request
   * @example
   * await client.social.friend(await client.social.resolve('halp'));
   * @returns false if the user is already friended, true otherwise
   * @throws {Error} If an error occurs (such as the user has blocked the client, etc)
   */ async friend(userID) {
        if (this.friends.find((r)=>r.id === userID)) return false;
        await this.#api.social.friend(userID);
        const userData = await this.who(userID);
        this.friends.push(new _relationship.Relationship({
            id: userData._id,
            relationshipID: "",
            username: userData.username,
            avatar: userData.avatar_revision
        }, this, this.#client));
        return true;
    }
    /**
   * Unfriend a user. Note: unfriending a user will unblock them if they are blocked.
   * @example
   * await client.social.unfriend(await client.social.resolve('halp'));
   * @returns false if the user is not unfriended, true otherwise
   */ async unfriend(userID) {
        if (!this.friends.find((r)=>r.id === userID)) return false;
        await this.#api.social.unfriend(userID);
        this.friends = this.friends.filter((r)=>r.id !== userID);
        return true;
    }
    /**
   * Block a user
   * @example
   * await client.social.block(await client.social.resolve('halp'));
   * @returns false if the user is already blocked, true otherwise
   */ async block(userID) {
        if (this.blocked.find((r)=>r.id === userID)) return false;
        await this.#api.social.block(userID);
    }
    /**
   * Unblock a user. Note: unblocking a user will unfriend them if they are friended.
   * @example
   * await client.social.unblock(await client.social.resolve('halp'));
   * @returns false if the user is not unblocked, true otherwise
   */ async unblock(userID) {
        await this.#api.social.unblock(userID);
    }
    /**
   * Invite a user to your room.
   * Note: the TETR.IO api doesn't send a response for a successful invite.
   * Instead, Triangle.js waits 100 milliseconds for an error, and resolves/returns if there is no error.
   * @example
   * await client.social.invite(await client.social.resolve('halp'));
   */ invite(userID) {
        this.#client.emit("social.invite", userID);
        return Promise.race([
            new Promise((_, reject)=>{
                const onError = (e)=>{
                    this.#client.off("client.error", onError);
                    reject(e);
                };
                this.#client.once("client.error", onError);
            }),
            new Promise((resolve)=>{
                setTimeout(resolve, 100);
            })
        ]);
    }
    /**
   * Set the client's status
   * @example
   * client.social.status('online', 'lobby:X-QP');
   */ // eslint-disable-next-line @typescript-eslint/no-wrapper-object-types
    status(status, detail = "") {
        this.#client.emit("social.presence", {
            status,
            detail
        });
    }
    /**
   * For internal use only. See `client.snapshot()`
   */ snapshot() {
        return {
            online: this.online,
            friends: this.friends.map((r)=>r.snapshot()),
            others: this.other.map((r)=>r.snapshot()),
            blocked: this.blocked.map((r)=>(0, _engine.deepCopy)(r)),
            config: (0, _engine.deepCopy)(this.config),
            notifications: (0, _engine.deepCopy)(this.notifications)
        };
    }
    /**
   * For internal use only. See `Client.fromSnapshot()`
   */ static fromSnapshot(snapshot, client) {
        return new Social(client, snapshot.config, snapshot);
    }
}
