// @haelp/teto/dist/utils/api/index.mjs
// export * as API from "./wrapper.mjs";
import { CONSTANTS } from "../constants.mjs";
import { channel } from "./channel.mjs";
import { core } from "./core.mjs";
import { relationship } from "./relationship.mjs";
import { rooms } from "./rooms.mjs";
import { server } from "./server.mjs";
import { users } from "./users.mjs";
export class API {
    defaults = {
        token: "",
        userAgent: CONSTANTS.userAgent,
        turnstile: null
    };
    get;
    post;
    rooms;
    server;
    users;
    social;
    channel;
    /** @hideconstructor */ constructor(options = {}){
        this.update(options);
    }
    update(options = {}) {
        Object.keys(options).forEach((key)=>{
            this.defaults[key] = options[key];
        });
        const b = core(this.defaults);
        this.get = b.get;
        this.post = b.post;
        this.rooms = rooms(this.get, this.post, this.defaults);
        this.server = server(this.get, this.post, this.defaults);
        this.users = users(this.get, this.post, this.defaults);
        this.social = relationship(this.get, this.post, this.defaults);
        this.channel = channel(this.get, this.post, this.defaults);
    }
}
export * as APITypes from "./wrapper.mjs";
