// @haelp/teto/dist/utils/api/index.js
// export * as API from "./wrapper";
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get API () {
        return API;
    },
    get APITypes () {
        return _wrapper;
    }
});
const _constants = require("../constants");
const _channel = require("./channel");
const _core = require("./core");
const _relationship = require("./relationship");
const _rooms = require("./rooms");
const _server = require("./server");
const _users = require("./users");
const _wrapper = /*#__PURE__*/ _interop_require_wildcard(require("./wrapper"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
class API {
    defaults = {
        token: "",
        userAgent: _constants.CONSTANTS.userAgent,
        turnstile: null
    };
    get;
    post;
    rooms;
    server;
    users;
    social;
    channel;
    /** @hideconstructor */ constructor(options = {}){
        this.update(options);
    }
    update(options = {}) {
        Object.keys(options).forEach((key)=>{
            this.defaults[key] = options[key];
        });
        const b = (0, _core.core)(this.defaults);
        this.get = b.get;
        this.post = b.post;
        this.rooms = (0, _rooms.rooms)(this.get, this.post, this.defaults);
        this.server = (0, _server.server)(this.get, this.post, this.defaults);
        this.users = (0, _users.users)(this.get, this.post, this.defaults);
        this.social = (0, _relationship.relationship)(this.get, this.post, this.defaults);
        this.channel = (0, _channel.channel)(this.get, this.post, this.defaults);
    }
}
