// @haelp/teto/dist/classes/game/player.mjs
import { Game } from "./index.mjs";
export var SpectatingState = /*#__PURE__*/ function(SpectatingState) {
    SpectatingState[SpectatingState["Inactive"] = 0] = "Inactive";
    SpectatingState[SpectatingState["Waiting"] = 1] = "Waiting";
    SpectatingState[SpectatingState["Active"] = 2] = "Active";
    return SpectatingState;
}({});
export class Player {
    name;
    gameid;
    userid;
    engine;
    /** Whether the player is currently receiving replay frames (actively spectated) */ state = 0;
    #client;
    #hook;
    #queue = [];
    #resolvers = [];
    #strategy;
    constructor(client, strategy, data, rawPlayers){
        this.name = data.options.username;
        this.gameid = data.gameid;
        this.userid = data.userid;
        this.engine = Game.createEngine(data.options, data.gameid, rawPlayers);
        this.#client = client;
        this.#hook = client.hook();
        this.#strategy = strategy;
        this.#hook.on("game.replay.state", ({ gameid, data })=>{
            if (gameid !== this.gameid) return;
            this.#resolvers.forEach(([r])=>r());
            this.#resolvers = [];
            this.state = 2;
            if (data === "early") {
            // todo: what to do here?
            } else if (data === "wait") {
            // todo: what to do here?
            } else {
                this.engine.fromSnapshot(Game.snapshotFromState(data.frame, this.engine.initializer, data.game));
            }
        });
        this.#hook.on("game.replay", ({ gameid, frames })=>{
            if (this.state !== 2) return;
            if (gameid !== this.gameid) return;
            if (this.engine.toppedOut) return;
            this.#queue.push(...frames);
        });
    }
    spectate() {
        return new Promise((resolve, reject)=>{
            if (this.state === 2) return;
            if (this.state === 1) return this.#resolvers.push([
                ()=>resolve(),
                (error)=>reject(error)
            ]);
            this.state = 1;
            this.#client.emit("game.scope.start", this.gameid);
            this.#resolvers.push([
                ()=>resolve(),
                (error)=>reject(error)
            ]);
        });
    }
    unspectate() {
        if (this.state === 0) return;
        this.#client.emit("game.scope.end", this.gameid);
        this.state = 0;
        this.#queue = [];
    }
    destroy() {
        this.#resolvers.forEach(([_, r])=>r("Game ended before spectating could begin"));
        if (this.state !== 0) this.unspectate();
        this.#resolvers = [];
        this.#hook.destroy();
        this.#queue = [];
        this.engine.events.removeAllListeners();
    }
    #tickOnce() {
        const frames = [];
        while(this.#queue.length > 0 && this.#queue[0].frame <= this.engine.frame){
            frames.push(this.#queue.shift());
        }
        this.engine.tick(frames);
    }
    /** @internal */ _tick() {
        if (this.state !== 2) return;
        if (this.#strategy === "instant") {
            while(this.#queue.some((f)=>f.frame > this.engine.frame)){
                this.#tickOnce();
            }
        } else if (this.#strategy === "smooth") {
            if (this.#queue.length === 0) return;
            const lastFrame = this.#queue[this.#queue.length - 1].frame;
            if (this.engine.frame < lastFrame - 20) {
                // if we're behind, skip ahead
                while(this.#queue.some((f)=>f.frame > this.engine.frame) && this.engine.frame < lastFrame - 20){
                    this.#tickOnce();
                }
            }
            if (this.#queue.some((f)=>f.frame > this.engine.frame)) {
                this.#tickOnce();
            }
        }
    }
    /** @internal */ set _strategy(strategy) {
        this.#strategy = strategy;
    }
    /** @internal */ get _strategy() {
        return this.#strategy;
    }
}
